### ISSUE TITLE
Facebook tracking parameter break url when shared link is clicked

### ISSUE THREAD

**Issue by julien-leclercq** [PrivateBin/PrivateBin#396]
## Disclaimer

I am not a php developper and haven't taken time to dive into the code, I searched unsuccessfully for similar issue but I maybe missed something. 
If the problem I am describing is unsolvable or already in dev pipe or has already been discussed and won't be solved be free to close this.

## Steps to reproduce
This has been encountered on https://www.privatebin.net and I haven't taken the time to try reproduce it neither in a local instance or an other public one

1.  create a paste and generate the link
2. share it on facebook 
3. click the link

### What happens
Facebook add a `fbclid` url parameter which break privatebin routing

### What should happen
In my humble opinion, the router should be able to remove such parameters or at least to ignore it since it is a common practice for social networks to add such tags for tracking purposes.

### Server address
http://www.privatebin.net

I can reproduce this issue on <https://privatebin.net>: Yes / No

**Comment by rugk** [comment]
If Facebook really does this, it rather looks like a problem of them… However, do you have an example for us? Example link? Possibly screenshot, if that helps?

**Comment by julien-leclercq** [comment]
this bin 
https://privatebin.net/?9fef56d4eaa605c4#gY3+KU9CcjmSi7Q/3Psf/jnKUOgIyevtPRv0NjHVAWY=

once shared on facebook it becomes this 
https://privatebin.net/?9fef56d4eaa605c4&fbclid=IwAR2uld0S-NQw4grmP5L2OvkU0Hfk4xjyW5ND8921OK6FStAlIlU2w91jagc#gY3+KU9CcjmSi7Q/3Psf/jnKUOgIyevtPRv0NjHVAWY= 

<img width="1174" alt="capture d ecran 2019-01-19 a 17 53 47" src="https://user-images.githubusercontent.com/9319408/51429836-49d28300-1c13-11e9-9392-f3fbbee84291.png">

Maybe you can decide that this is Facebook's problem. But, they are not the only one doing so and they won't change their way. I would totally understand that you do not want to fix for whatever reason or make it low prio. But I feel like you'll have issue with other referrers in the future if you keep it as it.

**Comment by julien-leclercq** [comment]
A good argument in favor of a fix imho is that users less competent with the way the web work could think there is an issue with their bin when there is not.

**Comment by rugk** [comment]
Okay, this is a really pesky thing. And wow, that identifier they include is longer than even our encryption key. :open_mouth: 
I personally agree, however, that it needs a fix. The fix would be a more flexible way how we parse our URL parameter.

----

So I searched for more information about this pesky thing and it seems Facebook introduced this [in October](https://stackoverflow.com/questions/52847475/what-is-fbclid-the-new-facebook-parameter) and they basically [just want to get around cookie/GDPR laws/tracking blockers](http://thisinterestsme.com/facebook-fbclid-parameter/).
They likely broke half the web by doing that stuff, but obviously they don't care…

---

Note for implementation: Could we possibly use the [new URL object](https://developer.mozilla.org/docs/Web/API/URL), at least where it is available (i.e. fallback if not)? This should be more fail-resistant than any RegEx or other workaround we might add and should work in case other companies decide to do weird stuff with URLs, too.

Some testing I did in the browser console:
```js
> const url = new URL("https://privatebin.net/?9fef56d4eaa605c4&fbclid=IwAR2uld0S-NQw4grmP5L2OvkU0Hfk4xjyW5ND8921OK6FStAlIlU2w91jagc#gY3+KU9CcjmSi7Q/3Psf/jnKUOgIyevtPRv0NjHVAWY=") 
> url.search // is not enough
"?9fef56d4eaa605c4&fbclid=IwAR2uld0S-NQw4grmP5L2OvkU0Hfk4xjyW5ND8921OK6FStAlIlU2w91jagc"
>  for (let p of url.searchParams) {
>  console.log(p);
>}
Array [ "9fef56d4eaa605c4", "" ]
Array [ "fbclid", "IwAR2uld0S-NQw4grmP5L2OvkU0Hfk4xjyW5ND8921OK6FStAlIlU2w91jagc" ]
```

Basically, we need to iterate all query strings and get the one that matches our format. Our format = the length of that string, is not it, @elrido? The length of the ID should always stay the same? If not, we may be getting into trouble identifying this properly…

**Alternative solution:** Implement a server redirect, but IMHO this would not be a nice workaround.

**Also:** One blog recommended that we may set a **canonical URL** (for SEO reasons, as otherwise these URLs with tracking IDs may be indexed). I doubt this makes much sense

**Comment by elrido** [comment]
@julien-leclercq Thank you for the example links!

- The [URL interface](https://developer.mozilla.org/docs/Web/API/URL) seems to be implemented differently across browsers and in jsDOM, that we use in node to emulate the browser environment for testing, it isn't supported yet, so we would have to use yet another library just to be able to use it in the tests. A solution evaluating the length of the string and optionally the types of characters seems to me simpler to implement and more robust then relying on more complex logic to deal with all the different browser implementations of that interface.
- A server redirect would remove the hash (not visible to the server) and so the decryption would fail. But it is correct that we have the same issue in PHP as well.

Let me add some unit tests for this problem (random additional query parameters thrown at the server or client) in a branch, so we can test any solutions against it. If I see any simple solution I'll add that to that branch, too, so you can evaluate it.

For reference, the (very simple) current logic to detect the paste IDs currently used can be found in these lines:
- [JS logic](https://github.com/PrivateBin/PrivateBin/blob/6bb599c81f03d9d33bcebc1968aca54c8d1abded/js/privatebin.js#L748) - takes the whole query string and removes the first letter (which is the question mark "?" at the beginning of the string)
- [PHP logic](https://github.com/PrivateBin/PrivateBin/blob/6bb599c81f03d9d33bcebc1968aca54c8d1abded/lib/Request.php#L103) - takes the whole query string as is (PHP strips the question mark for us)

**Comment by rugk** [comment]
Ugh this pesky jsDOM… Is it actually even implemented in _any_ way? I could not find anything in their doc or issues regarding this…
If jsDOM does not support modern stuff, I'd rather blame them and they may do in the future. We should not limit us from using proper modern technology just because some NodeJS lib does not support it.

>  so we would have to use yet another library just to be able to use it in the tests

…or we do not test these edge cases and we always use the "older browser" fallback, i.e. implementation we have now. I now, I also don't like this, but if jsdom is so stupid…

> same issue in PHP as well

Ugh, did not think about that. However, AFAIK, in PHP we also have a proper way of iterating query parameters, so we can find out our own.

---

The thing is, the PHP version is probably not the problem, but in JS we do have a _very robust_, modern and forward-compatible solution. To do some awkward hacks like evaluating "the types of characters" is hard there, I mean we can only check for Base64 chars, but the Facebook link may also contain them. And even the length is not to easy, because implementations may not always add their link at the end of the query string – what happens if they add it at the front?

So there is certainly some RegEx that can catch this, but it may then easily break

So as I already tested around with the URl interface, I could implement it this way. And I don't like not to use a proper (built-in) solution if we can, only because our testing lib does not support it. I mean, it's not that we would loose browser support, just our testing lib cannot do this … arggh…

**Comment by elrido** [comment]
- jsDOM: If you or anyone can provide us some other way to emulate the browser environment for our unit tests in node (which doesn't have a DOM or other browser APIs of any kind) or any other CLI based solution that is supported on travisCI and in docker, then please suggest it and consider rewriting 2000 lines of tests to the new solution. At the moment it is what we have, but I am willing to learn a new JS unit testing environment, if someone migrates us to it.
- I have pushed the tests to the [fb-tracking-param-fix branch](https://github.com/PrivateBin/PrivateBin/commit/cc53d95ed1f470a8143e8a56aa3192e2c4b3e13e) and will now start to work at at a quick-fix. I assume the fix will end up being be much smaller then the test cases for it.

**Comment by rugk** [comment]
So please create a PR… There we've got proper review tools and track it there, etc.

**Comment by julien-leclercq** [comment]
It maybe a dumb question but why didn't you made the router look like 
whateverprivatebinhost.org/bin/ENCRYPTEDKEY
or 
whateverprivatebinhost.org?key=ENCRYPTEDKEY
I am really ignorant with php but I do develop webapp and your routing strategy looks really weak to me

**Comment by rugk** [comment]
Apparently we got a regression here: https://github.com/PrivateBin/PrivateBin/issues/658

**Comment by Daxation** [comment]
change **URL** from: [](https://privatebin.net/?9fef56d4eaa605c4#gY3+KU9CcjmSi7Q/3Psf/jnKUOgIyevtPRv0NjHVAWY=)
to **URL encoded string**: [](https://examplebin.net/?9fef56d4eaa605c4%23gY3%2BKU9CcjmSi7Q%2F3Psf%2FjnKUOgIyevtPRv0NjHVAWY%3D)
using **Javascript** code:
`const url = new Url("https://privatebin.net/?" + encodeURIComponent("9fef56d4eaa605c4#gY3+KU9CcjmSi7Q/3Psf/jnKUOgIyevtPRv0NjHVAWY="));`

Now just change the page to ignore trying to process fbclid as any parameter of the paste and we are done.
